src/components/utils/test/timer_test.cc

```diff
@@ -43,214 +43,276 @@ namespace components {
 namespace timer_test {
 namespace {
 
-sync_primitives::Lock test_lock;
-sync_primitives::ConditionalVariable lock_monitor;
+sync_primitives::Lock shot_lock;
+sync_primitives::ConditionalVariable shot_condition;
+
+const std::string kTimerName = "TestTimer";
+const bool kSingleShot = true;
 const uint32_t kDefaultTimeout = 30u;
-const std::string kTimerName = "test_timer";
+const uint32_t kDefaultTimeoutRestart = 45u;
 
 class TestTask : public timer::TimerTask {
  public:
-  TestTask() : calls_count_(0u) {}
+  TestTask() : timer_(NULL), calls_count_(0u) {}
+
+  void set_timer(timer::Timer* timer) {
+    timer_ = timer;
+  }
+
+  virtual void PerformTimer() const {}
 
   void run() const OVERRIDE {
-    sync_primitives::AutoLock auto_lock(test_lock);
+    sync_primitives::AutoLock auto_lock(shot_lock);
     ++calls_count_;
-    lock_monitor.NotifyOne();
+    shot_condition.NotifyOne();
+    PerformTimer();
   }
 
-  uint GetCallsCount() const {
+  size_t calls_count() const {
     return calls_count_;
   }
 
- private:
-  mutable uint calls_count_;
+ protected:
+  mutable timer::Timer* timer_;
+  mutable size_t calls_count_;
 };
 
-class FakeClassWithTimer {
+class TestTaskWithStart : public TestTask {
  public:
-  FakeClassWithTimer()
-      : calls_count_(0u)
-      , internal_timer_("test_timer",
-                        new timer::TimerTaskImpl<FakeClassWithTimer>(
-                            this, &FakeClassWithTimer::OnTimer)) {
-    internal_timer_.Start(kDefaultTimeout, false);
+  void PerformTimer() const OVERRIDE {
+    if (timer_) {
+      timer_->Start(kDefaultTimeoutRestart, !kSingleShot);
+    }
   }
+};
 
-  void OnTimer() {
-    sync_primitives::AutoLock auto_lock_(test_lock);
-    internal_timer_.Stop();
-    ++calls_count_;
-    lock_monitor.NotifyOne();
+class TestTaskWithStop : public TestTask {
+ public:
+  void PerformTimer() const OVERRIDE {
+    if (timer_) {
+      timer_->Stop();
+    }
   }
+};
 
-  bool IsTimerRunning() const {
-    return internal_timer_.is_running();
-  }
+}  // namespace
 
-  uint GetCallsCount() const {
-    return calls_count_;
-  }
+// Start - Stop
 
- private:
-  uint calls_count_;
-  timer::Timer internal_timer_;
-};
-}  // namespace
+TEST(TimerTest, Start_Stop_NoLoop_NoCall) {
+  MockTimerTask* mock_task = new MockTimerTask();
```

Comment:
@malirod because we have to pass pointer to `Task` to `Timer` constructor. `Timer` will delete it in it's destructor.